Source/interactive_scene.cpp

```diff
@@ -1,270 +1,268 @@
-                    }
-                    else
-                    {
-                        LOGS_ERROR << "Unknown control type: " << control[RPC_CONTROL_KIND].as_string();
-                    }
-                }
-            }
-            catch (std::exception e)
-            {
-                LOGS_ERROR << "Failed to create scenes in process_get_scenes_reply";
-            }
-        }
-    }
-    catch (std::exception e)
-    {
-        LOGS_ERROR << "Failed in process_get_scenes_reply";
-    }
-
-    if (!success)
-    {
-        LOGS_ERROR << errorString;
-        // TODO: return error
-    }
-}
-
-interactive_scene_impl::interactive_scene_impl()
-{
-    m_interactivityManager = interactivity_manager::get_singleton_instance();
-}
-
-interactive_scene_impl::interactive_scene_impl(
-    _In_ string_t sceneId,
-    _In_ string_t etag,
-    _In_ bool disabled,
-    _In_ const std::vector<const string_t&>& controls
-) :
-    m_sceneId(std::move(sceneId)),
-    m_etag(std::move(etag)),
-    m_disabled(std::move(disabled))
-{
-    m_interactivityManager = interactivity_manager::get_singleton_instance();
-}
-
-NAMESPACE_MICROSOFT_MIXER_END
-
+//*********************************************************
+//
+// Copyright (c) Microsoft. All rights reserved.
+// THIS CODE IS PROVIDED *AS IS* WITHOUT WARRANTY OF
+// ANY KIND, EITHER EXPRESS OR IMPLIED, INCLUDING ANY
+// IMPLIED WARRANTIES OF FITNESS FOR A PARTICULAR
+// PURPOSE, MERCHANTABILITY, OR NON-INFRINGEMENT.
+//
+//*********************************************************
+#include "pch.h"
+#include "interactivity.h"
+#include "interactivity_internal.h"
+#include "json_helper.h"
+
+NAMESPACE_MICROSOFT_MIXER_BEGIN
+static std::mutex s_singletonLock;
+
+const string_t&
+interactive_scene::scene_id() const
+{
+    return m_impl->scene_id();
+}
+
+const std::vector<string_t>
+interactive_scene::groups()
+{
+    return m_impl->groups();
+}
+
+const std::vector<std::shared_ptr<interactive_button_control>>
+interactive_scene::buttons()
+{
+    return m_impl->buttons();
+}
+
+const std::shared_ptr<interactive_button_control>
+interactive_scene::button(_In_ const string_t& controlId)
+{
+    return m_impl->button(controlId);
+}
+
+const std::vector<std::shared_ptr<interactive_joystick_control>>
+interactive_scene::joysticks()
+{
+    return m_impl->joysticks();
+}
+
+const std::shared_ptr<interactive_joystick_control>
+interactive_scene::joystick(_In_ const string_t& controlId)
+{
+    return m_impl->joystick(controlId);
+}
+
+interactive_scene::interactive_scene() {
+    m_interactivityManager = interactivity_manager::get_singleton_instance();
+    m_impl = std::make_shared<interactive_scene_impl>();
+}
+
+interactive_scene::interactive_scene(
+    _In_ string_t sceneId,
+    _In_ bool enabled
+)
+{
+    m_interactivityManager = interactivity_manager::get_singleton_instance();
+    m_impl = std::make_shared<interactive_scene_impl>(sceneId, L"" /*etag filled in by service call*/, enabled);

```

Comment:
Yes we will. Might be a little while. There's some feature work on the roadmap we'll be doing first.

---
In reply to: [125970708](https://github.com/mixer/interactive-cpp/pull/17#discussion_r125970708) [](ancestors = 125970708)